### DISCUSSION TITLE
How do I make buttons interact with each other?

### DISCUSSION THREAD

**Discussion by MileyHollenberg** [elgatosf/elgato#20] [Help]
I'm working on a simple plugin to connect Clockify to my Stream Deck. The plan is to add a button per type of activity (my main game project, any current R&D project, admin stuff, etc..).

How would I interact with the other buttons when one is pressed? I want to make it so only one of them can be active at a time and instantly update the UI for that as well for all currently visible buttons of this plugin. I simply have 2 states for the button and it would need to be set back to the 0 state for all other buttons.

**Comment by ChekTek** [comment]
Hi!

The best way to handle this is by using the `willAppear` and `willDisappear` events to cache and delete the key information. You can create a Map of contexts to payloads, add the key in the `willAppear` event and remove the key in the `willDisappear` event. Then when a key is pressed, use the contexts you stored from the other keys to set the appropriate state.  (If you only need the context and no other information, you may want to use a Set instead of a Map)

https://developer.elgato.com/documentation/stream-deck/sdk/events-received/#willappear
https://developer.elgato.com/documentation/stream-deck/sdk/events-received/#willdisappear

**Comment by MileyHollenberg** [comment]
Hey there,

I've had a look at the willAppear event but I don't think this can solve my problem. The willAppear event only tells me about my current button and not the other buttons on the StreamDeck. I'd need to be able to access the other buttons and give them a sign to update their state but it feels like all the buttons are sandboxed so they can only affect themselves (which is totally understandable)

How would I setup communication between the buttons? I've currently got 3 on the same screen and individually they are working but I need them to be able to affect the others when one starts a time entry as that will automatically end the other trackers

**Comment by MileyHollenberg** [comment]
![image](https://github.com/elgatosf/elgato/assets/799917/2ff08d2d-8d6a-413f-bb97-d961c127e4e5)

Ignore my fabulous drawing skills :p, this is how they look right now. The top one is currently active and I want to be able to automatically reset it to inactive if I press either of the two buttons below it (regardless of their position obviously).

I'm currently using their local settings and state to keep track of whether they are active or not but I'm going to move that to the global settings to at least have the behavior linked up properly already but I'd need a way to also change the visual states of the other buttons so it's clear which time tracker is currently active.

**Comment by MileyHollenberg** [comment]
Deleted my previous 2 comments, basically I am not able to receive any updates to the other buttons when 1 changes. I thought I was but that was just all of them getting the willAppear event which makes sense for all of them.

**Comment by tiptronic** [comment]
Hi @MileyHollenberg 
was that a question - or does it work now?

**Comment by MileyHollenberg** [comment]
The question is basically, how do I communicate between the different buttons on my stream deck that a different button has taken over?

**Comment by tiptronic** [comment]
As Zack described earlier it's pretty easy:
- all your actions will receive an `willAppear` event
- you can use this event to cache the action information in a global variable `MYGLOBAL={}`  (e.g. `MYGLOBAL[context] = yourcurrentactionssettings`)
- now, using this `MYGLOBAL`, you know which actions are currently running and whatever you cached there
- on the `willDisappear` event you just remove the action from that global (e.g. `delete MYGLOBAL[context]`)
- That way you have access to your action's settings from any action, because the global can be accessed from every plugin's action.

**Comment by MileyHollenberg** [comment]
Ah I found my mistake, I totally forgot about the global and local states in Javascript and I was constantly overwriting a global state thinking it was a local one. Made some modifications and it now appears to be working as expected :). It's definitely not the cleanest code I've ever written but it's good enough for now.

I'm using the cache of context's to send a "getSettings" event to all of them to trigger them to update
